fdbclient/include/fdbclient/BlobGranuleRequest.actor.h

```diff
@@ -0,0 +1,186 @@
+				           blobGranuleMapping[i].key.printable(),
+				           blobGranuleMapping[i + 1].key.printable());
+			}
+			// no granule for range
+			throw blob_granule_transaction_too_old();
+		}
+
+		state UID workerId = decodeBlobGranuleMappingValue(blobGranuleMapping[i].value);
+		if (workerId == UID()) {
+			if (BGR_DEBUG) {
+				fmt::print("ERROR: Invalid Blob Worker ID for range [{0} - {1}) \n",
+				           blobGranuleMapping[i].key.printable(),
+				           blobGranuleMapping[i + 1].key.printable());
+			}
+			// no worker for granule
+			throw blob_granule_transaction_too_old();
+		}
+
+		if (!tr->trState->cx->blobWorker_interf.count(workerId)) {
+			Optional<Value> workerInterface = wait(tr->get(blobWorkerListKeyFor(workerId)));
+			// from the time the mapping was read from the db, the associated blob worker
+			// could have died and so its interface wouldn't be present as part of the blobWorkerList
+			// we persist in the db.
+			if (workerInterface.present()) {
+				tr->trState->cx->blobWorker_interf[workerId] = decodeBlobWorkerListValue(workerInterface.get());
+			} else {
+				if (BGR_DEBUG) {
+					fmt::print("ERROR: Worker  for range [{1} - {2}) does not exist!\n",
+					           workerId.toString().substr(0, 5),
+					           blobGranuleMapping[i].key.printable(),
+					           blobGranuleMapping[i + 1].key.printable());
+				}
+				break;
+			}
+		}
+
+		if (BGR_DEBUG) {
+			fmt::print("Requesting range [{0} - {1}) from worker {2}!\n",
+			           blobGranuleMapping[i].key.printable(),
+			           blobGranuleMapping[i + 1].key.printable(),
+			           workerId.toString().substr(0, 5));
+		}
+
+		KeyRangeRef range(blobGranuleMapping[i].key, blobGranuleMapping[i + 1].key);
+		request.reply.reset();
+		request.setRange(range);
+		// TODO consolidate?
+		BlobWorkerInterface bwi = tr->trState->cx->blobWorker_interf[workerId];
+		RequestStream<Request, P> const* stream = &(bwi.*channel);
+		Future<ErrorOr<REPLY_TYPE(Request)>> response = stream->tryGetReply(request);
+		requests.push_back(response);
+	}
+	// wait for each request. If it has an error, retry from there if it is a retriable error
+	state int j = 0;
+	for (; j < requests.size(); j++) {
+		try {
+			ErrorOr<REPLY_TYPE(Request)> result = wait(requests[j]);
+			if (result.isError()) {
+				throw result.getError();
+			}
+			results.push_back(results.arena(), result.get());
+		} catch (Error& e) {
+			if (e.code() == error_code_wrong_shard_server || e.code() == error_code_request_maybe_delivered ||
+			    e.code() == error_code_broken_promise || e.code() == error_code_connection_failed) {
+				// re-read mapping and retry from failed req
+				i = j;
+				break;
+			} else {
+				if (BGR_DEBUG) {
+					fmt::print("ERROR: Error doing request for range [{0} - {1}): {2}!\n",
+					           blobGranuleMapping[j].key.printable(),
+					           blobGranuleMapping[j + 1].key.printable(),
+					           e.name());
+				}
+				throw;
```

Comment:
do we need a Trace for error case?

Reply:
We could, my concern is that it'd be spammy. I could always do it suppressed.